src/core.ts

```diff
@@ -180,85 +180,92 @@ const patchOuter = patchFactory((node, fn, data) => {
 /**
  * Checks whether or not the current node matches the specified nameOrCtor and
  * key.
- *
  * @param matchNode A node to match the data to.
  * @param nameOrCtor The name or constructor to check for.
- * @param key An optional key that identifies a node.
- * @param typeId An type identifier that avoids reuse between elements that
- *     would otherwise match.
+ * @param key The key used to identify the Node.
  * @return True if the node matches, false otherwise.
  */
 function matches(
-    matchNode: Node, nameOrCtor: NameOrCtorDef, key: string|null|undefined,
-    typeId: {}|null|undefined) {
+    matchNode: Node, nameOrCtor: NameOrCtorDef, key: Key) {
   const data = getData(matchNode);
 
   // Key check is done using double equals as we want to treat a null key the
   // same as undefined. This should be okay as the only values allowed are
   // strings, null and undefined so the == semantics are not too weird.
-  // tslint:disable:triple-equals
-  return nameOrCtor == data.nameOrCtor && typeId == data.typeId &&
-      key == data.key;
-  // tslint:enable:triple-equals
+  // tslint:disable-next-line:triple-equals
+  return nameOrCtor == data.nameOrCtor && key == data.key;
 }
 
 
 /**
- * Aligns the virtual Node definition with the actual DOM, moving the
- * corresponding DOM node to the correct location or creating it if necessary.
+ * Finds the matching node, starting at `node` and looking at the subsequent
+ * siblings if a key is used.
+ * @param node The node to start looking at.
  * @param nameOrCtor The name or constructor for the Node.
- * @param key The key used to identify the Node..
- * @param typeId An type identifier that avoids reuse between elements that
- *     would otherwise match.
+ * @param key The key used to identify the Node.
  */
-function alignWithDOM(
-    nameOrCtor: NameOrCtorDef, key: string|null|undefined,
-    typeId: {}|null|undefined) {
-  if ((currentNode && matches(currentNode, nameOrCtor, key, typeId))) {
-    return;
+function getMatchingNode(
+    matchNode: Node|null, nameOrCtor: NameOrCtorDef, key: Key) {
+  if (!matchNode) {
+    return null;
+  }
+
+  if (matches(matchNode, nameOrCtor, key)) {
+    return matchNode;
+  }
+
+  if (!key) {
+    return null;
   }
 
-  const parent = currentParent!;
-  const parentData = getData(parent);
-  const keyMap = parentData.keyMap;
+  return getMatchingNode(matchNode.nextSibling, nameOrCtor, key);
```

Comment:
Nit: How about unrolling this into a while loop. Safari is the only browser to implement tail call recursion. Chrome and the others will have to trampoline through returns.

Reply:
Hmm, yeah probably better to not use a tail call here. Will take a look.